[UnifiedPDF] <embed> and <object> PDFs spill out of their frames when printed

aprotyas opened the pull request:
#### 0f96a6a34a86e64caae90cb4f0bc5349811cba76
<pre>
[UnifiedPDF] &lt;embed&gt; and &lt;object&gt; PDFs spill out of their frames when printed
<a href="https://bugs.webkit.org/show_bug.cgi?id=270253">https://bugs.webkit.org/show_bug.cgi?id=270253</a>
<a href="https://rdar.apple.com/122500785">rdar://122500785</a>

Reviewed by Tim Horton.

We never clip the GraphicsContext to be painted in with the clip rect we
pass to paintPDFContent, which causes the PDF content to spill over the
frame border. This patch fixes that.

Note that this is not a problem for non-software painting paths because
when we vend painting responsibilites to PDFKit, they perform the
clipping correctly.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::paintPDFContent):

Canonical link: <a href="https://commits.webkit.org/275487@main">https://commits.webkit.org/275487@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5eb3bace0cc7e6fee907b5060cd2f2d8c2763957

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44386 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44578 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38104 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24183 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18338 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17938 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36161 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45998 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38182 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37519 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16808 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18427 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9404 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18487 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -705,6 +705,8 @@ static String mutationObserverNotificationString()
     auto pageCoverage = pageCoverageForRect(clipRect);
     auto documentScale = pageCoverage.pdfDocumentScale;
 
+    context.clip(clipRect);
```
I would do this in the caller. We don't need to clip here in the normal layer painting code path, because the GraphicsLayer code has already applied a clip.

Also, for printing, you should avoid having the tiles paint cached buffers while printing, so that needs wiring up,.

aprotyas replied:
Okay, I'll change where we clip.

aprotyas replied:
> Also, for printing, you should avoid having the tiles paint cached buffers while printing, so that needs wiring up.

Yes! I'll make a bugzilla for that.

aprotyas committed:
[UnifiedPDF] <embed> and <object> PDFs spill out of their frames when printed
https://bugs.webkit.org/show_bug.cgi?id=270253
rdar://122500785

Reviewed by Tim Horton.

We never clip the GraphicsContext to be painted in with the clip rect we
pass to paintPDFContent, which causes the PDF content to spill over the
frame border. This patch fixes that.

Note that this is not a problem for non-software painting paths because
when we vend painting responsibilites to PDFKit, they perform the
clipping correctly.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::paintPDFContent):

Canonical link: https://commits.webkit.org/275487@main
